pkg/virt-api/validating-webhook/validating-webhook.go

```diff
@@ -0,0 +1,406 @@
+	for idx, disk := range disks {
+		// Verify only a single device type is set.
+		deviceTargetSetCount := 0
+		if disk.Disk != nil {
+			deviceTargetSetCount++
+		}
+		if disk.LUN != nil {
+			deviceTargetSetCount++
+		}
+		if disk.Floppy != nil {
+			deviceTargetSetCount++
+		}
+		if disk.CDRom != nil {
+			deviceTargetSetCount++
+		}
+
+		// NOTE: not setting a device target is okay. We default to Disk.
+		// However, only a single device target is allowed to be set at a time.
+		if deviceTargetSetCount > 1 {
+			errors = append(errors, fmt.Errorf("spec.domain.devices.disks[%d] can only have a single target type defined", idx))
+		}
+
+	}
+
+	return errors
+}
+
+func validateVolumes(volumes []v1.Volume) []error {
+	errors := []error{}
+
+	for idx, volume := range volumes {
+		// Verify exactly one source is set
+		volumeSourceSetCount := 0
+		if volume.PersistentVolumeClaim != nil {
+			volumeSourceSetCount++
+		}
+		if volume.CloudInitNoCloud != nil {
+			volumeSourceSetCount++
+		}
+		if volume.RegistryDisk != nil {
+			volumeSourceSetCount++
+		}
+		if volume.Ephemeral != nil {
+			volumeSourceSetCount++
+		}
+		if volume.EmptyDisk != nil {
+			volumeSourceSetCount++
+		}
+
+		if volumeSourceSetCount != 1 {
+			errors = append(errors, fmt.Errorf("spec.volumes[%d] must have exactly one source type set", idx))
+		}
+
+		// Verify cloud init data is within size limits
+		if volume.CloudInitNoCloud != nil {
+			noCloud := volume.CloudInitNoCloud
+			userDataLen := 0
+
+			userDataSourceCount := 0
+			if noCloud.UserDataSecretRef != nil && noCloud.UserDataSecretRef.Name != "" {
+				userDataSourceCount++
+			}
+			if noCloud.UserDataBase64 != "" {
+				userDataSourceCount++
+				userData, err := base64.StdEncoding.DecodeString(noCloud.UserDataBase64)
+				if err != nil {
+					errors = append(errors, fmt.Errorf("spec.volumes[%d].cloudInitNoCloud.userDataBase64 is not a valid base64 value.", idx))
+				}
+				userDataLen = len(userData)
+			}
+			if noCloud.UserData != "" {
+				userDataSourceCount++
+				userDataLen = len(noCloud.UserData)
+			}
+
+			if userDataSourceCount != 1 {
+				errors = append(errors, fmt.Errorf("spec.volumes[%d].cloudInitNoCloud must have one exactly one userdata source set.", idx))
+			}
+
+			if userDataLen > cloudInitMaxLen {
+				errors = append(errors, fmt.Errorf("spec.volumes[%d].cloudInitNoCloud userdata exceeds %d byte limit", idx, cloudInitMaxLen))
+			}
+		}
+	}
+	return errors
+}
+
+func validateDevices(devices *v1.Devices) []error {
+	errors := []error{}
+	errors = append(errors, validateDisks(devices.Disks)...)
+	return errors
+}
+
+func validateDomainSpec(spec *v1.DomainSpec) []error {
+	errors := []error{}
+	errors = append(errors, validateDevices(&spec.Devices)...)
+	return errors
+}
+
+func validateVirtualMachineSpec(spec *v1.VirtualMachineSpec) []error {
+	errors := []error{}
+
+	// Validate disks and VolumeNames match up correctly
+	for idx, disk := range spec.Domain.Devices.Disks {
```

Comment:
* Could you also check if a volume is only referenced once by a disk?
 * Could you also check that the names of the volumes are unique?

 I would prefer n instead of n*n and adding a  limit to the array (lets start with 255?). For this and the additional checks, maps may help.

Reply:
done